back/spec/services/content_image_service_spec.rb

```diff
@@ -0,0 +1,213 @@
+require 'rails_helper'
+
+describe ContentImageService do
+  before do
+    @text_image_1, @text_image_2 = create_list :text_image, 2
+    allow(TextImage).to receive(:create!).and_return(@text_image_1, @text_image_2)
+  end
+
+  let(:subclass) do
+    Class.new(described_class) do
+      def decode_content(str)
+        JSON.parse str, symbolize_names: true
+      rescue StandardError => e
+        raise ContentImageService::DecodingError.new parse_errors: e.message
+      end
+
+      def encode_content(json)
+        json.to_json
+      end
+
+      def image_elements(content)
+        content.select { |elt| elt[:type] == 'image' }
+      end
+
+      def content_image_class
+        TextImage
+      end
+
+      def image_attributes(img_elt, imageable, field)
+        { imageable: imageable, imageable_field: field, remote_image_url: img_elt[image_attribute_for_element] }
+      end
+
+      def code_attribute_for_element
+        :data_cl2_content_image_code
+      end
+
+      def image_attribute_for_element
+        :src
+      end
+
+      def code_attribute_for_model
+        :text_reference
+      end
+
+      def precompute_for_rendering(_multiloc, _imageable, _field)
+        @precomputed_content_images = content_image_class.all.index_by do |ti|
+          ti[code_attribute_for_model]
+        end
+      end
+
+      def fetch_content_image(code)
+        @precomputed_content_images[code]
+      end
+
+      def content_image_url(content_image)
+        "/images/#{content_image[code_attribute_for_model]}.jpg"
+      end
+    end
+  end
+  let(:service) { subclass.new }
+
+  describe 'swap_data_images' do
+    it 'returns exactly the same input locales' do
+      imageable = build(:user, bio_multiloc: { 'en' => '[]', 'fr-BE' => '[]', 'nl-BE' => '[]', 'de' => '[]' })
+      expect(TextImage).not_to receive :create!
```

Comment:
When I see the `not_to receive :something` pattern in a test, I am always wondering the usefulness, because I could write `expect(TextImage).not_to receive :bogus` and the test would pass.
It is clear that no new `TextImage` should be created, but probably it is not important to verify that `create!` is not sent. Probably it is beter to verify that no new `TextImage` records are created in the database. Then the test can verify that no side effect happened, instead of verifying that a message was not sent.

Reply:
The alternative I think of is to take the counts of TextImage before and after and compare them. But I think you once said that you were not fond of that approach either? How would you test if new records have been created?

Reply:
> The alternative I think of is to take the counts of TextImage before and after and compare them. But I think you once said that you were not fond of that approach either?

It think it is the way to go here.